### PR TITLE
adding 3d mesh tallies

### PR DESCRIPTION
**Pull request by Shimwell**
## Proposed changes

This PR will add the ability for users to add 3D mesh tallies to the simulation

This first commit is just a small refactoring of the existing cell and mesh tally logic

Not sure if the code from the workshop for getting mesh data into vtk files should be added or if it should run as an external script
https://github.com/ukaea/openmc_workshop/blob/1a4abd540c0bb18b5ac6115a7a127892fd776d97/tasks/task_08_CSG_mesh_tally/openmc-statepoint-3d.py#L53-L162

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Code refactoring
- [ ] Documentation Update (if none of the other choices apply)
- [ ] New tests

## Checklist

- [ ] Pep8 applied
- [ ] Unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by Shimwell** [comment]
The refactoring has reduced the number of lines and therefore the coverage percent has dropped a bit.

File: paramak/parametric_neutronics/neutronics_model.py [hunk 1]
```diff
@@ -175,6 +179,28 @@ def mesh_tally_2D(self, value):
...
 
+    @property
+    def mesh_tally_3D(self):
+        return self._mesh_tally_3D
+
+    @mesh_tally_3D.setter
+    def mesh_tally_3D(self, value):
+        if value is not None:
+            if not isinstance(value, list):
+                raise ValueError(
```
Comment (RemDelaporteMathurin) [inline paramak/parametric_neutronics/neutronics_model.py:190]: Shouldn't this be a TypeError ? :-D

File: paramak/parametric_neutronics/neutronics_model.py [hunk 2]
```diff
@@ -372,129 +398,128 @@ def create_neutronics_model(self, method=None):
...
+            mesh_xyz.upper_right = [
+                self.geometry.largest_dimension,
+                self.geometry.largest_dimension,
+                self.geometry.largest_dimension
+            ]
+
+            for standard_tally in self.mesh_tally_3D:
+                if standard_tally == 'tritium_production':
+                    score = '(n,Xt)'  # where X is a wild card
+                    prefix = 'tritium_production'
```
Comment (RemDelaporteMathurin) [inline paramak/parametric_neutronics/neutronics_model.py:419]: When standard_tally is tritium_production, these lines aren't tested

File: paramak/parametric_neutronics/neutronics_model.py (hunk 1 above)
Comment (Shimwell) [reply paramak/parametric_neutronics/neutronics_model.py:190]: ah you caught me copying and pasting from the 2d mesh @property and forgot to implement the pullrequest review suggestions

Comment (Shimwell) [reply paramak/parametric_neutronics/neutronics_model.py]: I fixed a few more of these in https://github.com/ukaea/paramak/commit/f9f7667838209ff043c2a6e0c618774e12faed3a

**Comment by Shimwell** [comment]
This is starting to work and can produce vtk files automatically :tada: 

Screenshot from 2020-12-23 20-04-15

**Comment by Shimwell** [comment]
I am updating the dependencies image so that it contains a pip install vtk command

**Comment by Shimwell** [comment]
I am not quite sure why the latest Circle CI tests are failing. It appears to be due to the Sweep tests.

The dependency image was recently updated so perhaps that is the cause

**Comment by RemDelaporteMathurin** [comment]
@Shimwell are the tests failing locally ?

**Comment by Shimwell** [comment]
> @Shimwell are the tests failing locally ?

Just checked this and yes all the tests pass locally

**Comment by Shimwell** [comment]
Circle CI tests for the sweep are still failing, 

This is strange because:
- This PR has no changes to the Sweep functions
- The tests pass locally
- The docker image for circle CI has been updated so that it is now very similar to the docker image built during the Github Actions tests (which pass)

**Comment by RemDelaporteMathurin** [comment]
> Circle CI tests for the sweep are still failing, 
> 
> 
> 
> This is strange because:
> 
> - This PR has no changes to the Sweep functions
> 
> - The tests pass locally
> 
> - The docker image for circle CI has been updated so that it is now very similar to the docker image built during the Github Actions tests (which pass)

Are you sure the image you are using for local testing is the same as the one used by the CI ?

**Comment by Shimwell** [comment]
ah sorry when I mentioned the tests pass locally it is not on a docker image but it is on my base system. I can try locally with the docker image. Thanks for the idea

**Comment by Shimwell** [comment]
The plot thickens as I just noticed that PR #648 is failing the same sweep tests. This is again happening with the Circle CI but not with Github Actions CI.

**Pull request merged**